File: lib/tui/frame.rb
# frozen_string_literal: true

require_relative "ansi"
require_relative "theme"
require_relative "screen_layout"

module Tui
  # Pure frame builder: given content, returns an array of strings, one per
  # terminal row, each exactly `width` visible characters. No IO — the app
  # decides how to paint, tests can assert on the result.
  module Frame
    A = Ansi
    T = Theme

    module_function

    # rows:     array of Views::Row
    # selected: row index to highlight (or nil)
    # footer:   array of interior lines; the symbol :rule draws a divider
    # popup:    { lines: [...], row: Integer, col: Integer } overlaid on body
    # panel:    { title:, lines: [...] } drawn in a fixed-width right pane
    # modal:    { title:, lines: [...] } drawn as a centered box over the body
    def build(width:, height:, header:, rows:, selected: nil, footer: [], popup: nil, panel: nil, modal: nil,
              layout: nil, panel_offset: 0)
      layout ||= ScreenLayout.new(width: width, height: height, footer: footer, selected: selected,
                                  panel: !panel.nil?, panel_offset: panel_offset)
      width = layout.width
      height = layout.height
      w = width - 2
      footer = layout.footer
      body_h = layout.body_height
      offset = layout.viewport_offset

      list_w = layout.list_width
      body = (rows[offset, body_h] || []).map.with_index do |row, vi|
        if vi == layout.selected_screen_row
          selected_row(row, list_w)
        else
          "  " + row.text
        end
      end
      body.map! { |line| A.vpad(A.vtrunc(line, list_w), list_w) }
      # Empty filler rows are pre-padded so we don't pay a second full-body
      # vpad pass just to expand a short viewport.
      body.fill(" " * list_w, body.size...body_h)

      render_panel!(body, panel, layout) if panel

      # Modal first, then the popup on top. Archive confirmation and forms must
      # remain visible above a persistent task-detail panel.
      overlay_modal!(body, modal, w - 2) if modal
      overlay!(body, popup, w - 2) if popup

      lines = []
      lines << "┌#{"─" * w}┐"
      lines << "│#{A.vpad(A.vtrunc(header, w), w)}│"
      lines << "├#{"─" * w}┤"
      body.each { |b| lines << "│ #{A.vpad(b, w - 2)} │" }
      lines << "├#{"─" * w}┤"
      footer.each do |f|
        lines << (f == :rule ? "├#{"─" * w}┤" : "│#{A.vpad(A.vtrunc(f, w), w)}│")
      end
      lines << "└#{"─" * w}┘"
      lines
    end

    def render_panel!(body, panel, layout)
      content_width = layout.panel_content_width
      panel_lines = [T.paint(:panel_title, A.vtrunc(panel[:title], content_width))]
      panel_lines << T.paint(:muted, "─" * content_width)
      panel_lines.concat(panel[:lines])

      body.map!.with_index do |list_line, index|
        content = A.vpad(A.vtrunc(panel_lines[index].to_s, content_width), content_width)
        "#{A.vpad(A.vtrunc(list_line, layout.list_width), layout.list_width)}│ #{content}"
      end
    end

    # The selection cursor glyph. Deliberately distinct from Views::MARK_COLLAPSED
    # ("▸ ") so a selected collapsed row reads "❯ ▸ title", not a doubled marker.
    # One cell + a trailing space = two cells, matching the marker column width.
    CURSOR = "❯ "

    # Render the selected row with its own field colors intact ON TOP of the
    # :selection background, via SGR compositing (no stripping/repaint).
    #
    # Contract: the row text already carries its field SGRs, each closed with a
    # reset (\e[0m — the only reset form Ansi.color emits). We (1) open the line
    # with the :selection SGR, (2) re-open it immediately after every reset so a
    # field's own fg/attrs layer over the selection background instead of
    # clearing it, (3) pad the visible text to the full inner width so the
    # background spans the row, then (4) close with a single reset. Truncation
    # runs FIRST, on the composed cursor+text, so the pad+reset tail can never be
    # clipped. A :selection that resolves to nothing (an unstyled theme) skips
    # compositing and just pads the plain cursor+text.
    def selected_row(row, w)
      body = A.vtrunc(CURSOR + row.text, w)
      sel = T.sgr(:selection)
      return A.vpad(body, w) if sel.empty?

      composited = A.composite(sel, body)
      pad = w - A.vislen(body)
      composited += " " * pad if pad.positive?
      composited + "\e[0m"
    end

    # Paste popup lines over the body starting at terminal-cell coordinates,
    # preserving styled base content on either side. Cell slices replace any
    # partially covered wide grapheme with padding, so content after the popup
    # remains in its original column.
    def overlay!(body, popup, w)
      popup[:lines].each_with_index do |pl, k|
        r = popup[:row] + k
        next if r.negative? || r >= body.size

        col = popup[:col].to_i
        source_start = [-col, 0].max
        dest_start = [col, 0].max
        next if dest_start >= w

        base = A.vpad(A.vtrunc(body[r], w), w)
        replacement = A.cell_slice(pl, source_start, w - dest_start)
        replacement_width = A.vislen(replacement)
        prefix = A.vpad(A.cell_slice(base, 0, dest_start), dest_start)
        suffix_start = dest_start + replacement_width
        suffix = A.cell_slice(base, suffix_start, w - suffix_start)
        body[r] = A.vpad(A.vtrunc(prefix + replacement + suffix, w), w)
      end
    end

    # Box up modal content and center it over the body via overlay!.
    # modal[:width] pins the box width (Modal computes it from the full
    # content so scrolling can't resize the box); without it the width
    # fits the visible lines. The title strip is painted with the
    # :modal_title theme slot, so themes can give it a background.
    def overlay_modal!(body, modal, w)
      if body.size < 3 || w < 4
        compact = [T.paint(:modal_title, modal[:title]), *modal[:lines]].first(body.size)
        overlay!(body, { lines: compact, row: 0, col: 0 }, w)
        return
      end

      bw = modal[:width] ||
           [(modal[:lines].map { |l| A.vislen(l) }.max || 0), A.vislen(modal[:title]) + 6, 30].max + 4
      bw = [[bw, w].min, 4].max
      inner = modal[:lines].map { |l| A.vtrunc(l, bw - 4) }
      inner = inner.first(body.size - 2)
      title = A.vtrunc(" #{modal[:title]} ", bw - 4)
      box = ["┌─#{T.paint(:modal_title, title)}#{"─" * [bw - 4 - A.vislen(title), 0].max}─┐"]
      inner.each { |l| box << "│ #{A.vpad(l, bw - 4)} │" }
      box << "└#{"─" * (bw - 2)}┘"
      overlay!(body, {
        lines: box,
        row: modal.fetch(:row, [(body.size - box.size) / 2, 0].max),
        col: modal.fetch(:col, [(w - bw) / 2, 0].max),
      }, w)
    end
  end
end

File: internal/tui/term/frame/frame.go
// Package frame is a pure frame builder: given content, it returns one string
// per terminal row, each exactly `width` visible cells wide. No IO — the caller
// decides how to paint, and tests assert on the result.
//
// Go port of Ruby's lib/tui/frame.rb.
package frame

import (
	"strings"

	"github.com/marcus/tasks/internal/tui/term/ansi"
	"github.com/marcus/tasks/internal/tui/term/border"
	"github.com/marcus/tasks/internal/tui/term/layout"
	"github.com/marcus/tasks/internal/tui/term/theme"
)

// Row is one list row. Text carries the row's own field styling, each field
// closed with a reset — the compositing contract selectedRow depends on.
type Row struct{ Text string }

// Panel is the fixed-width right pane.
type Panel struct {
	Title string
	Lines []string
}

// Options describes one frame.
type Options struct {
	Width, Height int
	Header        string
	Rows          []Row
	// Selected is the absolute row index to highlight, or nil.
	Selected  *int
	Footer    []layout.FooterLine
	Popup     *layout.Popup
	Panel     *Panel
	Modal     *layout.Modal
	Layout    *layout.Layout
	Theme     *theme.Theme
	Truecolor bool
	// PanelOffset is only consulted when Layout is nil.
	PanelOffset int
}

// Cursor is the selection glyph. Deliberately distinct from the collapsed-row
// marker ("▸ ") so a selected collapsed row reads "❯ ▸ title", not a doubled
// marker. One cell plus a trailing space = two cells, matching the marker
// column width.
const Cursor = "❯ "

// Build renders the whole frame.
func Build(opts Options) []string {
	th := opts.Theme
	if th == nil {
		th = theme.Default()
	}
	lay := opts.Layout
	if lay == nil {
		lay = layout.New(layout.Options{
			Width: opts.Width, Height: opts.Height, Footer: opts.Footer,
			Selected: opts.Selected, Panel: opts.Panel != nil,
			PanelOffset: opts.PanelOffset,
		})
	}
	width := lay.Width
	height := lay.Height
	w := width - 2
	footer := lay.Footer
	bodyH := lay.BodyHeight
	listW := lay.ListWidth

	visible := layout.VisibleRows(lay, opts.Rows)
	body := make([]string, 0, bodyH)
	for vi, row := range visible {
		if lay.SelectedScreenRow != nil && vi == *lay.SelectedScreenRow {
			body = append(body, selectedRow(th, row, listW))
		} else {
			body = append(body, "  "+row.Text)
		}
	}
	for i := range body {
		body[i] = ansi.VPad(ansi.VTrunc(body[i], listW), listW)
	}
	// Empty filler rows are pre-padded so we do not pay a second full-body pad
	// pass just to expand a short viewport.
	filler := strings.Repeat(" ", listW)
	for len(body) < bodyH {
		body = append(body, filler)
	}

	if opts.Panel != nil {
		renderPanel(th, body, *opts.Panel, lay)
	}

	// Modal first, then the popup on top. Archive confirmation and forms must
	// remain visible above a persistent task-detail panel.
	if opts.Modal != nil {
		overlayModal(th, body, *opts.Modal, w-2, opts.Truecolor)
	}
	if opts.Popup != nil {
		overlay(body, *opts.Popup, w-2)
	}

	// The outer ring is UNDRAWN, and so are the rules that used to sit above
	// and below the body. Four rows and two columns went to saying "the
	// application is here", which the alternate screen already says, and the
	// rules cut the list and the detail rail beside it into stacked boxes.
	// Blank rows separate them now. See internal/tui/render.go, whose output
	// this must match cell for cell.
	//
	// The border painter and its gradient are NOT retired with the ring: every
	// floating surface — modal, popup, the editor form — still draws a box, and
	// a box there is doing real work, separating the surface from what is
	// behind it. Only the frame that had nothing behind it is gone.
	blank := strings.Repeat(" ", width)
	pad := func(text string) string { return " " + ansi.VPad(ansi.VTrunc(text, w), w) + " " }

	lines := make([]string, 0, height)
	lines = append(lines, pad(opts.Header))
	lines = append(lines, blank)
	for _, b := range body {
		lines = append(lines, "  "+ansi.VPad(b, w-2)+"  ")
	}
	lines = append(lines, blank)
	for _, f := range footer {
		if f.Rule {
			lines = append(lines, blank)
			continue
		}
		lines = append(lines, pad(f.Text))
	}
	return lines
}

func renderPanel(th *theme.Theme, body []string, panel Panel, lay *layout.Layout) {
	contentWidth := lay.PanelContentWidth
	panelLines := []string{
		th.Paint("panel_title", ansi.VTrunc(panel.Title, contentWidth)),
		th.Paint("muted", strings.Repeat("─", contentWidth)),
	}
	panelLines = append(panelLines, panel.Lines...)

	for index := range body {
		source := ""
		if index < len(panelLines) {
			source = panelLines[index]
		}
		content := ansi.VPad(ansi.VTrunc(source, contentWidth), contentWidth)
		listLine := ansi.VPad(ansi.VTrunc(body[index], lay.ListWidth), lay.ListWidth)
		body[index] = listLine + "│ " + content
	}
}

// selectedRow renders the selected row with its own field colors intact ON TOP
// of the selection background, via SGR compositing (no stripping or repaint).
//
// Contract: the row text already carries its field SGRs, each closed with a
// reset. We (1) open the line with the selection SGR, (2) re-open it
// immediately after every reset so a field's own fg/attrs layer over the
// selection background instead of clearing it, (3) pad the visible text to the
// full inner width so the background spans the row, then (4) close with a
// single reset. Truncation runs FIRST, on the composed cursor+text, so the
// pad+reset tail can never be clipped. A selection slot that resolves to
// nothing (an unstyled theme) skips compositing and just pads.
func selectedRow(th *theme.Theme, row Row, w int) string {
	body := ansi.VTrunc(Cursor+row.Text, w)
	sel := th.SGR("selection")
	if sel == "" {
		return ansi.VPad(body, w)
	}
	composited := ansi.Composite(sel, body)
	if pad := w - ansi.VisLen(body); pad > 0 {
		composited += strings.Repeat(" ", pad)
	}
	return composited + "\x1b[0m"
}

// overlay pastes popup lines over the body starting at terminal-cell
// coordinates, preserving styled base content on either side. Cell slices
// replace any partially covered wide grapheme with padding, so content after
// the popup remains in its original column.
func overlay(body []string, popup layout.Popup, w int) {
	for k, pl := range popup.Lines {
		r := popup.Row + k
		if r < 0 || r >= len(body) {
			continue
		}
		col := popup.Col
		sourceStart := 0
		if col < 0 {
			sourceStart = -col
		}
		destStart := col
		if destStart < 0 {
			destStart = 0
		}
		if destStart >= w {
			continue
		}

		base := ansi.VPad(ansi.VTrunc(body[r], w), w)
		replacement := ansi.CellSlice(pl, sourceStart, w-destStart)
		replacementWidth := ansi.VisLen(replacement)
		prefix := ansi.VPad(ansi.CellSlice(base, 0, destStart), destStart)
		suffixStart := destStart + replacementWidth
		suffix := ansi.CellSlice(base, suffixStart, w-suffixStart)
		body[r] = ansi.VPad(ansi.VTrunc(prefix+replacement+suffix, w), w)
	}
}

// overlayModal boxes up modal content and centers it over the body via overlay.
// Modal.Width pins the box width (so scrolling cannot resize the box); without
// it the width fits the visible lines. The title strip is painted with the
// modal_title slot, so themes can give it a background.
func overlayModal(th *theme.Theme, body []string, modal layout.Modal, w int, truecolor bool) {
	if len(body) < 3 || w < 4 {
		compact := append([]string{th.Paint("modal_title", modal.Title)}, modal.Lines...)
		if len(compact) > len(body) {
			compact = compact[:len(body)]
		}
		overlay(body, layout.Popup{Lines: compact, Row: 0, Col: 0}, w)
		return
	}

	bw := modal.Width
	if bw == 0 {
		widest := 0
		for _, l := range modal.Lines {
			if v := ansi.VisLen(l); v > widest {
				widest = v
			}
		}
		bw = maxInt(maxInt(widest, ansi.VisLen(modal.Title)+6), 30) + 4
	}
	bw = maxInt(minInt(bw, w), 4)

	inner := make([]string, 0, len(modal.Lines))
	for _, l := range modal.Lines {
		inner = append(inner, ansi.VTrunc(l, bw-4))
	}
	if len(inner) > len(body)-2 {
		inner = inner[:len(body)-2]
	}
	boxed := make([]string, 0, len(inner))
	for _, l := range inner {
		boxed = append(boxed, " "+ansi.VPad(l, bw-4)+" ")
	}
	title := th.Paint("modal_title", ansi.VTrunc(" "+modal.Title+" ", bw-4))
	box := border.Box(border.BoxOptions{
		InnerLines: boxed,
		InnerWidth: bw - 2,
		Gradient:   th.Gradient("border"),
		Solid:      th.SGR("border"),
		Truecolor:  truecolor,
		Title:      title,
		TitleLead:  1,
	})

	row := maxInt((len(body)-len(box))/2, 0)
	col := maxInt((w-bw)/2, 0)
	if modal.Placed {
		row, col = modal.Row, modal.Col
	}
	overlay(body, layout.Popup{Lines: box, Row: row, Col: col}, w)
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}
